Initial implementation of CyHy configuration library

felddy opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

This is the initial implementation of the CyHy configuration tool.

## 💭 Motivation and context ##

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

`cyhy-config` simplifies, and standardizes the way we locate, parse, and validate configuration files in both the filesystem and in AWS SSM.  

## 🧪 Testing ##

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- These boxes should remain unchecked until the pull request has been -->
<!-- approved. -->

- [x] Revert dependencies to default branches.
- [x] Finalize version.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [x] Create a release.

felddy committed:
First commit before transfer to cisagov

Just committing what I have so that we can move it over to
cisagov for collaboration.

felddy committed:
Auto install types dependencies for mypy

felddy committed:
Remove unused debug flag from cyhy_config.py

felddy committed:
Sort imports

felddy committed:
Add py.typed file to allow users to type check

felddy committed:
Reconfigure logging

felddy committed:
Remove unused imports

felddy committed:
Stop logging to root logger

felddy committed:
Allow passing model to configuration reader

felddy committed:
Remove cyhy_config.py

felddy committed:
Drop support for Python 3.11

Log filtering only works in 3.12

felddy committed:
Add boto3 library

felddy committed:
Replace find and read with get function

felddy committed:
Add support for reading from SSM

Consolidate read_config.py into this file.

felddy committed:
Rename find_config to cyhy_config

felddy committed:
Remove unused models

Models are now defined by the caller.

felddy committed:
Fix blank lines for linter

felddy committed:
Remove entrypoint

felddy committed:
Lazy instantiation of SSM client

felddy committed:
Add documentation to README

felddy committed:
Remove unused test configuration and database entries

felddy committed:
Rename pytest config

felddy committed:
Retarget coverage to correct source

felddy committed:
Update pytest configuration to include debug logging

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

felddy committed:
Add new tests for cyhy_config

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

felddy committed:
Add comment regarding PEP 561

Stolen from mypy:
https://github.com/python/mypy/blob/9ffb9ddf80fbf92956a55946c8f023fc09638959/mypy/py.typed

felddy committed:
Update build from skeleton and limit Python to ≥3.11

jsf9k commented on setup-env:
```diff
@@ -251,7 +251,7 @@ for req_file in "requirements-dev.txt" "requirements-test.txt" "requirements.txt
 
 # Install all necessary mypy type stubs
-mypy --install-types src/
+mypy --install-types --non-interactive src/
```
Can this change be added to [cisagov/skeleton-generic](https://github.com/cisagov/skeleton-generic)?  Also, does the command return a zero return code with the addition of this flag?  (This is a problem I ran into just this morning.)

I'm happy to create the PR if folks agree.

felddy replied:
I agree this should be pushed up to the skeleton. 

I confirmed  command does return with exit `0` by printing the output in `setup-env`. 

```console
...
Installing missing stub packages:
~/.pyenv/versions/cyhy-config/bin/python3 -m pip install types-requests

Collecting types-requests
  Downloading types_requests-2.32.0.20240914-py3-none-any.whl.metadata (1.9 kB)
Requirement already satisfied: urllib3>=2 in ~/.pyenv/versions/cyhy-config/lib/python3.12/site-packages (from types-requests) (2.2.3)
Downloading types_requests-2.32.0.20240914-py3-none-any.whl (15 kB)
Installing collected packages: types-requests
Successfully installed types-requests-2.32.0.20240914

Success: no issues found in 3 source files
mypy exit code: 0
...
```

jsf9k replied:
This code actually comes from [cisagov/skeleton-python-library](https://github.com/cisagov/skeleton-python-library) (not [cisagov/skeleton-generic](https://github.com/cisagov/skeleton-generic) as I thought) so I created cisagov/skeleton-python-library#138.

felddy committed:
Refactor logger to be under the common CyHy namespace

jsf9k commented on setup.py:
```diff
@@ -75,25 +75,27 @@ def get_version(version_file):
     include_package_data=True,
-    install_requires=["docopt", "schema", "setuptools >= 24.2.0"],
+    install_requires=[
+        "boto3",
+        "pydantic",
+        "rich",
+        "setuptools >= 24.2.0",
```
The stubs libraries for these dependencies [should be added to `.pre-commit-config.yaml`](https://github.com/cisagov/cyhy-config/blob/5624d094e448da3e34b838fc49325e8866913ab9/.pre-commit-config.yaml#L158-L160) if they exist.

Note that [`pydantic` is a little funky](https://github.com/pydantic/pydantic/discussions/9446#discussioncomment-9615466) in this respect.

Also, the `types-docopt` entry in `.pre-commit-config.yaml` should be removed since it is no longer needed.

felddy replied:
See: 
- a0daf8a4db445fa2ef2e36a6840df102dfc24d90
- 11b3bff54a54f8661feb5b0e83f544b577e8748d

jsf9k replied:
Please also remove the `types-docopt` entry in `.pre-commit-config.yaml` as that dependency is no longer being used.

jsf9k replied:
Ah, nevermind.  I see that you did it now in commit 9183bbf.

felddy committed:
Full test coverage

Move scenario comments closer to implementation.
Fix bugs found in testing.

felddy committed:
Separate scenarios into individual tests

felddy committed:
Remove unused dependency

felddy committed:
Add missing dependencies for mypy hook

felddy committed:
Remove unused hook dependency types-docopt

dav3r commented on pytest.ini:
```diff
@@ -1,2 +1,2 @@
 [pytest]
-addopts = -v -ra --cov
+addopts = -v -ra --cov --log-cli-level=DEBUG
```
Just a heads-up that https://github.com/cisagov/skeleton-python-library/pull/139 is a thang now.

dav3r commented on src/cyhy_config/cyhy_config.py:
```diff
@@ -0,0 +1,183 @@
+
+This module provides a function to find a CyHy configuration file.
+
+It will search in the following locations:
+ - A location passed in as a parameter
+ - Environment variables for an SSM parameter path
+ - Environment variables for a file path
+ - The current working directory
+ - The user's home directory
+ - The system's /etc directory
```
Minor verbiage tweaks:
```suggestion
It will search in the following locations, in this order:
 - A location passed in as a parameter
 - Environment variable for an SSM parameter path
 - Environment variable for a file path
 - The current working directory
 - The user's home directory
 - The system's /etc directory
```

jsf9k replied:
In addition, I think that this should be an ordered list, e.g.:
```
1. A location passed in as a parameter
1. Environment variable for an SSM parameter path
```
and so on.

felddy replied:
See:
- c88c22b6c5b4c8dc69b4cb8501bf0f2f0494b206

dav3r replied:
LGTM, thanks.

dav3r commented on src/cyhy_config/cyhy_config.py:
```diff
@@ -0,0 +1,183 @@
+            )
+
+    # Check the current working directory
+    cwd_path = Path("cyhy.toml")
```
Should `cyhy.toml` be defined as a constant since we use it in several places here?

felddy replied:
See:
- [c88c22b](https://github.com/cisagov/cyhy-config/pull/5/commits/c88c22b6c5b4c8dc69b4cb8501bf0f2f0494b206)

felddy replied:
See:
- 927e78747d801bd846dfaacdb2dee043554b5cbc

dav3r commented on tests/test_cyhy_config.py:
```diff
@@ -0,0 +1,213 @@
+
+
+def test_read_config_ssm_other_client_error():
+    """Test read_config_ssm.
+
+    When SSM responds with an error code other than ParameterNotFound.
+    """
```
Make this consistent with the other descriptions:
```suggestion
    """Test read_config_ssm when SSM responds with an error code other than ParameterNotFound."""
```

felddy replied:
See:
- e59ede9e36fbeb1075906e040787cdd8a900fee0

dav3r reviewed (approved):
Approved, though I noted a few minor thangs.

Strong work on this and bonus points for full code test coverage! 🦾 💼

felddy committed:
Add missing release test

felddy committed:
Add function overload signatures to appease mypy

felddy committed:
Fix incorrect version import

felddy committed:
Reenable coverage pragma

When we define custom exclude_lines for coverage, the existing pragma is cleared out.

See: https://coverage.readthedocs.io/en/latest/config.html#report-exclude-lines

felddy committed:
Exclude overload passes from coverage

felddy committed:
Improve search procedure description

felddy committed:
Normalize comment to improve readability

felddy committed:
Rename test model as the name was misleading

felddy committed:
Convert Paths to constants

Updated tests to use new constants.
Renamed environment variable constants for clarity.

jsf9k commented on setup.py:
```diff
@@ -75,25 +75,26 @@ def get_version(version_file):
+    ],
     extras_require={
         "test": [
+            "boto3-stubs",
```
Shouldn't the other stub dependencies listed in `.pre-commit-config.yaml` be listed here as well?

jsf9k replied:
@cisagov/vm-dev - This is something that may need to added to [cisagov/skeleton-python-library](https://github.com/cisagov/skeleton-python-library) as well.  Should we do it in `test` or in a new `dev` section?

felddy replied:
See: bc903bbe142afbed2080abf756cce5501a06ac93

jsf9k replied:
Commit bc903bb is good enough for now, but I'd still like to get a consensus on [this question](https://github.com/cisagov/cyhy-config/pull/5#discussion_r1775341759) so I can create an upstream PR.

dav3r replied:
If the stubs are needed for pre-commit mypy tests to pass, then `test` seems appropriate to me.  I can see the case for `dev` as well, but I don't think it makes a huge difference one way or the other.

felddy replied:
Sorry I missed responding to that question.  I agree that the type requirements should be separated out.

See:
- 572b7f1ad9989f2553444885e7fa01ca4e8d5263

jsf9k replied:
I created cisagov/skeleton-python-library#141 to address this upstream, but the `setup-env` script still does not run successfully there even with the changes from cisagov/skeleton-python-library#138.  Any ideas what needs to be done there?

felddy committed:
Remove outdated comment and dependency pin on Coveralls

The version restriction for Coveralls was added in February 2020 due to a
specific issue with v1.11.0. However, Coveralls is now at v4.0.1, and the
previous concern is no longer relevant.

In the future, if we need to temporarily embargo a version, we should document
the reason in an issue and set a reminder to remove the embargo once the problem
is resolved.

felddy committed:
Update pre-commit configuration and setup.py dependencies

Replace proxy package `types-boto3` with `boto3-stubs` for consistency.

felddy committed:
Remove undocumented restriction on setuptools

Removing reference to version of setuptools from July of 2016: https://pypi.org/project/setuptools/24.2.0/

felddy committed:
Take responsibility for authorship

felddy committed:
Create dev dependency section in setup.py

We are now separating out the dev dependencies from the test dependencies.

Added a comment in the pre-commit configuration similar to the one in setup.py
to remind developers to keep dependencies in sync

dav3r reviewed (approved):
👍 👍

felddy committed:
Add comment about pragma line

felddy committed:
Update version file path in bump_version.sh

felddy committed:
Bump version from 0.1.0 to 1.0.0
